source/common/network/connection_impl.cc

```diff
@@ -797,7 +842,17 @@ void ConnectionImpl::onWriteReady() {
       }
     } else {
       ASSERT(bothSidesHalfClosed() || delayed_close_state_ == DelayedCloseState::CloseAfterFlush);
-      closeConnectionImmediately();
+      ENVOY_CONN_LOG(trace, "both sides are half closed or it is final close after flush state",
+                     *this);
+      if (delayed_close_state_ == DelayedCloseState::CloseAfterFlush) {
+        // close() is already managed by the filter manager and delayed.
+        // This is the final close.
+        closeConnectionImmediately();
+      } else if (bothSidesHalfClosed()) {
+        // If half_close is enabled, the close should still go through the filter manager, since
+        // the read is possible still pending.
+        closeThroughFilterManager({ConnectionEvent::LocalClose, true});
+      }
```

Comment:
What the filter chain can do if the both ways are closed? I think we cannot write or read with this socket?

Reply:
Yes, but we need to let the data in the flight in the filter manager continue. Here it is possible that the write finishes, but read is still propagating the `end_stream` to next filters.